Major Updates (#98)
tgdwyer/tgdwyer.github.io: _chapters/functionaljavascript.md

For many standard loops, however, the logic is the same every time and can easily be abstracted into a function.  Examples: Array.map, Array.reduce, Array.forEach, etc.  The logic of the loop body is specified with a function which can execute in its own scope, without the risk of breaking the loop logic.

#### Examples

1. Consider this loop to multiply each item in the `someArray` by two. Try to find the mistake in this code: 

    ```javascript
    const someArray = [1,2,3,4,5];
    let newArray = [];
    for (let i = 0; i <= someArray.length; i++) {
        newArray.push_back(someArray[i] * 2);
    }
    ```

    <p class="spoiler">The condition should be `<` not `<=`</p>

    To avoid the likelihood of errors we can replace the `for` loop with the use of `.map`

    We use the `map` function, since we want to apply a function to every element in the list

    ```javascript
    const someArray = [1,2,3,4,5];
    const newArray = someArray.map(x => x * 2);
    ``` 

    We use an arrow function here to allow our function definition to be short and to the point! 

> 2. Consider this function which aims to sum up a list.  Try to find the mistake in this code: 

Reviewer: ```suggestion
2. Consider this code which aims to compute the product of a list. Try to find the mistake in this code:
```

    ```javascript
    const someArray = [1,2,3,4,5];
    let result = 1;
    for (let i = 0; i < someArray.length; i++) {
        result *= i
    }
    ```
    <p class="spoiler">We should multiply by `someArray[i]` not `i`</p>

>     Again, to avoid the likelihood of errors we can replace the `for` loop with the use of `.reduce`

Reviewer: ```suggestion
    Again, to avoid the likelihood of errors, we can replace the `for` loop with the use of `.reduce`.
```

For many standard loops, however, the logic is the same every time and can easily be abstracted into a function.  Examples: Array.map, Array.reduce, Array.forEach, etc.  The logic of the loop body is specified with a function which can execute in its own scope, without the risk of breaking the loop logic.

#### Examples

1. Consider this loop to multiply each item in the `someArray` by two. Try to find the mistake in this code:

    ```javascript
    const someArray = [1,2,3,4,5];
    let newArray = [];
    for (let i = 0; i <= someArray.length; i++) {

>         newArray.push_back(someArray[i] * 2);

Reviewer: ```suggestion
        newArray.push(someArray[i] * 2);
```

        {name: "Bob", age: 15},
        {name: "Charlie", age: 30},
        {name: "David", age: 10},
    ];
    let firstChildName = undefined;
    for (let i = 0; i < people.length; i++) {
        if (people[i].age < 18) {
            firstChildName = people[i].name;
            break;
        }
    }
    ```

4. Refactor this code to remove the loop:

    ```javascript
    const people = [
        {name: "Alice", age: 20},
        {name: "Bob", age: 15},
        {name: "Charlie", age: 30},
        {name: "David", age: 10},
    ];
    let result = "";
    for (let i = 0; i < people.length; i++) {
        result += `Person #${i + 1}: ${people[i].name} is ${people[i].age} years old`;
        if (i != people.length - 1) {
            result += "\n";
        }
    }
    ```

#### Solutions

1. We can use `filter` to get the even numbers, then `map` to divide each of those numbers by 2:

    ```javascript
    const numbers = [2, 6, 3, 7, 10];
    const result = numbers.filter(x => x % 2 === 0).map(x => x / 2);
    ```

2. We need to sum up the lengths of each word. One way to do that would be to use `reduce`:

    ```javascript
    const words = ["apple", "banana", "cherry"];

>     const totalLength = words.reduce((acc, x) => acc + x.length);

Author: please comment on accumulator and why you can it be missing

Author: ignore my grammar i got distracted by shiny object half way through the sentence

Reviewer: oops, fixed
